In [367]:
# Import packages

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import root_mean_squared_error
from sklearn.metrics import r2_score

pd.set_option('display.max_columns', None)

In [368]:
# Load the dataset (Seattle Airbnb)

df_raw = pd.read_csv("listings.csv.gz")
df = df_raw.copy()

In [369]:
df.columns

Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
       'description', 'neighborhood_overview', 'picture_url', 'host_id',
       'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
       'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
       'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'ca

In [377]:
df["neighbourhood_group_cleansed"].value_counts()

neighbourhood_group_cleansed
Other neighborhoods    1371
Downtown                899
Central Area            637
Capitol Hill            585
Queen Anne              576
West Seattle            503
Ballard                 417
Rainier Valley          381
Cascade                 289
University District     287
Beacon Hill             285
Northgate               238
Delridge                218
Lake City               123
Magnolia                111
Seward Park              59
Interbay                 17
Name: count, dtype: int64

In [370]:
df.describe()

,id,scrape_id,host_id,host_listings_count,host_total_listings_count,latitude,longitude,accommodates,bathrooms,bedrooms,beds,minimum_nights,maximum_nights,minimum_minimum_nights,maximum_minimum_nights,minimum_maximum_nights,maximum_maximum_nights,minimum_nights_avg_ntm,maximum_nights_avg_ntm,calendar_updated,availability_30,availability_60,availability_90,availability_365,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,availability_eoy,number_of_reviews_ly,estimated_occupancy_l365d,estimated_revenue_l365d,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
count,6.996000e+03,6.996000e+03,6.996000e+03,6645.000000,6645.000000,6996.000000,6996.000000,6996.000000,6288.000000,6845.000000,6285.000000,6996.000000,6996.000000,6980.000000,6980.000000,6980.000000,6980.000000,6996.000000,6996.000000,0.0,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6996.000000,6.221000e+03,6109.000000,6109.000000,6109.000000,6109.000000,6109.000000,6109.000000,6109.000000,6996.000000,6996.000000,6996.000000,6996.000000,6109.000000
mean,6.685861e+17,2.025093e+13,1.735477e+08,42.310609,86.618360,47.626181,-122.334414,4.010292,1.456266,1.661066,2.157677,10.990709,389.699257,10.349140,26.242980,641.738825,679.903295,11.723228,658.415695,NaN,13.083047,32.436249,52.898513,207.709405,82.307604,16.992138,1.489565,57.852058,16.201830,105.077330,1.873634e+04,4.808836,4.837687,4.816284,4.879180,4.883393,4.834426,4.717764,28.839909,25.622927,3.074328,0.029017,2.226317
std,5.597414e+17,0.000000e+00,1.930086e+08,189.624035,505.638438,0.049222,0.033363,2.595572,0.834621,1.182109,1.519486,18.718235,429.112014,17.873247,70.109523,491.383994,479.818142,18.884772,481.099525,NaN,10.541648,20.947358,30.820507,122.928557,126.641425,21.979380,2.101167,33.246350,24.657595,93.147852,3.545103e+04,0.338030,0.301480,0.334917,0.255026,0.290954,0.262944,0.357484,76.091530,75.410953,10.971933,0.443056,2.116236
min,6.606000e+03,2.025093e+13,2.536000e+03,1.000000,1.000000,47.498030,-122.419478,1.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,NaN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.010000
25%,4.268168e+07,2.025093e+13,2.135475e+07,1.000000,1.000000,47.601869,-122.356998,2.000000,1.000000,1.000000,1.000000,1.000000,30.000000,1.000000,2.000000,100.000000,364.000000,2.000000,180.225000,NaN,2.000000,14.000000,29.000000,90.000000,5.000000,1.000000,0.000000,32.000000,0.000000,6.000000,2.520000e+03,4.780000,4.820000,4.780000,4.870000,4.890000,4.800000,4.670000,1.000000,1.000000,0.000000,0.000000,0.630000
50%,7.646666e+17,2.025093e+13,9.075747e+07,3.000000,4.000000,47.624260,-122.332635,4.000000,1.000000,1.000000,2.000000,2.000000,365.000000,2.000000,3.000000,365.000000,1000.000000,3.000000,571.100000,NaN,13.000000,37.000000,62.000000,234.000000,32.000000,9.000000,1.000000,67.000000,5.000000,86.000000,1.333800e+04,4.900000,4.920000,4.910000,4.950000,4.970000,4.900000,4.810000,2.000000,2.000000,0.000000,0.000000,1.840000
75%,1.166228e+18,2.025093e+13,3.080316e+08,15.000000,21.000000,47.663476,-122.311914,5.000000,2.000000,2.000000,3.000000,30.000000,365.000000,30.000000,30.000000,1125.000000,1125.000000,30.000000,1125.000000,NaN,22.000000,51.000000,79.000000,327.000000,105.000000,27.000000,2.000000,85.250000,26.000000,182.000000,2.730000e+04,4.980000,4.980000,4.980000,5.000000,5.000000,4.970000,4.900000,11.000000,7.000000,0.000000,0.000000,3.290000
max,1.516826e+18,2.025093e+13,7.186610e+08,2989.000000,8403.0000

In [371]:
df["minimum_nights"].value_counts()

minimum_nights
2      2033
1      1894
30     1469
3       702
31      324
4       137
5       100
7        66
12       66
6        36
60       27
28       22
90       21
180      18
14       14
10       12
21        6
15        6
100       4
120       4
19        4
80        3
40        3
85        3
300       2
45        2
150       2
32        1
365       1
27        1
26        1
75        1
140       1
87        1
351       1
65        1
25        1
99        1
9         1
13        1
50        1
170       1
200       1
Name: count, dtype: int64

In [ ]:
# Data Preprocessing : Convert string type to integer or boolean

## price
df["price"] = pd.to_numeric(
    df["price"].astype(str).str.replace(r"[\$,]", "", regex=True)
)

## host_response_time
mapping_host_response_time = {
    'within an hour': 1,
    'within a few hours': 2,
    'within a day': 3,
    'a few days or more': 4
}

df['host_response_time'] = df['host_response_time'].map(mapping_host_response_time)

## target variable (review_scores_rating) : high score (1) vs. low score (0) (threshold: 4.91 (50% quartile))
threshold = 4.91

df['review_scores_class'] = (df["review_scores_rating"] > threshold).astype(int)

In [ ]:
# Feature Engineering

## add relative pricing
group = ["bedrooms", "bathrooms", "neighbourhood_cleansed", "room_type"]
df["avg_price_group"] = df.groupby(group)["price"].transform("mean")
df["relative_price"] = df["price"] / df["avg_price_group"]

## minimum nights -> short-term listing vs. long-term listing
threshold = 10

df['is_long_term'] = (df["minimum_nights"] > threshold).astype(int)

In [ ]:
df["minimum_nights"].value_counts()

In [ ]:

# Filter uninterested columns

cols_interested = [
    "relative_price",
#    "price",
    "room_type",
    "bedrooms",
    "bathrooms",
    "beds",
    "accommodates",
#    "neighbourhood_cleansed",
    "neighbourhood_group_cleansed",
    "is_long_term",
    "host_response_time",
    "host_identity_verified",
    "number_of_reviews",
    "estimated_occupancy_l365d",
    "estimated_revenue_l365d",
#    "number_of_reviews_ltm",
#    "number_of_reviews_l30d",
    "review_scores_class"      # target variable
]

df = df[cols_interested]
df = df.dropna()
df.info()

In [ ]:
df.describe()

* Column Description

0 relative_price : Mean is 0.98 and median is 0.97. Most listings are priced close to comparable listings. The maximum value is 7.39, indicating possible outliers.

1 bedrooms : Maximum is 13, minimum is 0, and median is 1. Most listings have 1–2 bedrooms.

2 bathrooms : Maximum is 16, minimum is 0, and median is 1. Most listings have one bathroom.

3 beds : Maximum is 14, minimum is 0, and median is 2.

4 accommodates : Maximum is 16, minimum is 1, and mean is 4.17.

5 minimum_nights : Maximum is 351 and minimum is 1. Very large values may represent long-term stays.

6 host_response_time : Mean is 1.17, indicating most hosts respond quickly.

7 number_of_reviews : Maximum is 1577 and minimum is 1.

8 review_scores_class : Higher scores group is 49.4%, and lower scores group is 50.6%.

In [ ]:
# run RF model

X = df.drop(columns=["review_scores_class"])
y = df["review_scores_class"]

X = pd.get_dummies(X, drop_first=True)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

rf = RandomForestClassifier(
    n_estimators=200,
    random_state=42
)

rf.fit(X_train, y_train)

In [ ]:
y_pred = rf.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print("Accuracy:", accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

In [ ]:
importance = pd.Series(
    rf.feature_importances_,
    index=X_train.columns
).sort_values(ascending=False)

importance.head(15)

In [ ]:
importance.head(10).plot(kind="barh")
plt.title("Feature importance for Seattle Airbnb rating prediction")
plt.show()

In [ ]:
from sklearn.inspection import PartialDependenceDisplay

X_test = X_test.astype(float)

features = [
    "estimated_revenue_l365d",
    "estimated_occupancy_l365d",
    "number_of_reviews",
    "relative_price",
    "accommodates",
    "beds",
    "bathrooms",
    "bedrooms"
]

fig, ax = plt.subplots(figsize=(12,8))

PartialDependenceDisplay.from_estimator(
    rf,
    X_test,
    features,
    target=1,
    ax=ax
)

plt.tight_layout()
plt.show()

In [ ]:
from sklearn.inspection import permutation_importance

result = permutation_importance(rf, X_test, y_test)

importance = pd.Series(result.importances_mean, index=X_test.columns)
importance.sort_values(ascending=False)